### ISSUE TITLE
joining channels issues

### ISSUE THREAD

**Issue by willkg** [mozilla/standup-irc#42]
I'm in the process of nixing rogerroger for the standups irc bot. In doing that, I copied down the list of channels that rogerroger was in to the standups db then restarted the standups bot. standups would join 26 channels, then hang and restart 4 minutes later. This continued for some time until either an irc op fixed the issue on their end or something else happened or something.

Anyhow, I suspect there's some throttling for joining a bunch of channels or something going on. It'd be good to know what it is and how we can adjust standup-irc to be a better citizen.

Some thoughts:

Can we join a bunch of channels at the same time in one call or do we have to do it one at a time?

Is this a problem just with joining a bunch of channels it wasn't previously in? Like maybe you can only join 20 new channels at a time?

Would it help to space out joining channels so at startup it joins one channel per second or something like that?

**Comment by pmac** [comment]
I'll bet our friend @fox2mike would know some of these answers.

**Comment by fox2mike** [comment]
Or knows someone who does! Calling @initrd for help

**Comment by initrd** [comment]
Thanks for looping in!

> 18:38 < levin.mozilla.org> ANNOUNCEMENT: User rogerroger RecvQ of 10240 exceeds connect class maximum of 4600

Documentation about [RecvQ](https://wiki.inspircd.org/Recvq). This can be fixed in a few ways:

1. The bot could stagger joins
2. You could try setting up the bot with AJOIN (/msg NickServ help ajoin). Not 100% sure this would help but at least the bot wouldn't be sending a bunch of /join at the same time.
3. Worst case, I could bump the recvq threshold for the IP. This is the least preferred option since the IP will need to be tracked if/every time it changes.

**Comment by willkg** [comment]
@initrd That helps a ton! We shouldn't change the recvq threshold for the ip since it's on heroku and who knows whether the ip changes or not. Probably does.

I'll change the code to stagger joins.

**Comment by willkg** [comment]
@initrd, @fox2mike Also, if there's anything else that rogerroger/standups irc bots are doing that's dumb/unhelpful/etc, please let me know. Writing IRC bots is not something I have a lot of experience with, so I don't know what I don't know.

**Comment by initrd** [comment]
Sure!

1. Please register the bot's nick (/msg Nickserv help register)!
2. Some channels (Mozilla IRC and elsewhere) have flood controls. It would be good if the bot were aware of the limits, and stagger multi-line messages.

**Comment by mozfreddyb** [comment]
@initrd The RecvQ (and SendQ) are per connection class. Do you think it would be possible to assign a different connection class to friendly bots?

This would simplify the bot and help servicing the users without unexpected delays (that would cause flooding/repetition on the user's end). The bot owners would of course promise they do not abuse the privilege.

This would also help solving #43.

**Comment by initrd** [comment]
Sure, I can take look at [configs](https://wiki.inspircd.org/Configuration#.3Cconnect.3E) and see what can be done. However, I can't promise a timeline, sorry!

**Comment by willkg** [comment]
I don't think we need to make any ircd config changes. I don't think there's much urgency here--rogerroger ran for years and this has only come up now.

The critical part of this issue that I couldn't do was understand what the actual problem was and what the options were. Now that we know what the problem is and we know it's not critical, we can fix it in our leisure.

Right now I'm focusing on getting standup v2 out the door. Once that happens, I can think more about this.

**Comment by pmac** [comment]
I've registered the nicks for `standups` and `standupstage` and configured the bots to identify. Hopefully that'll help. And as I had to restart the bots to do that it's interesting to note that `standups` had no problem rejoining channels. 👍
